pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,160 @@
+
+	var outBuilder, errBuilder strings.Builder
+	cmd.Stdout = &outBuilder
+	cmd.Stderr = &errBuilder
+
+	e.log.Info("fence agent start", "uid", uid, "command", command)
+
+	var err error
+	for i := 0; i < 5; i++ {
+		err = cmd.Run()
+		if err == nil {
+			break
+		}
+		e.log.Error(err, "command failed", "uid", uid, "command", command)
+		time.Sleep(1 * time.Second)
 	}
 
-	// Execute the Post request for SPDY (shell) executor
-	err = execSPDY.Stream(remotecommand.StreamOptions{
-		Stdout: &stdoutBuf,
-		Stderr: &stderrBuf,
-		Tty:    false,
-	})
-	if err != nil {
-		e.log.Error(err, "Failed to run exec command", "stdout", stdoutBuf.String(), "stderr", stderrBuf.String())
+	e.log.Info("fence agent done", "uid", uid, "command", command, "stdout", outBuilder.String(), "stderr", errBuilder.String(), "err", err)
+
+	// loop to handle only the cases where
+	// - FAR cannot be found, but it does exist
+	// - the status update fails for conflict
+	for {
+		far, err := e.getFARByUID(uid)
+		if err != nil {
+			if !apiErrors.IsNotFound(err) {
+				continue
+			}
+			e.log.Error(err, "could not update status", "FAR uid", uid)
+			break
+		}
+
+		err = e.updateStatus(far, err)
+		if err == nil {
+			break
+		}
+		if apiErrors.IsConflict(err) {
+			e.log.Error(err, "conflict while updating the status", "FAR uid", uid)
+			time.Sleep(1 * time.Second)
+		}
+		e.log.Error(err, "failed to update status", "FAR uid", uid)
+		break
+	}
+}
+
+func (e *executer) getFARByUID(uid types.UID) (*v1alpha1.FenceAgentsRemediation, error) {
+	farList := &v1alpha1.FenceAgentsRemediationList{}
+	if err := e.List(e.ctx, farList, &client.ListOptions{}); err != nil || len(farList.Items) == 0 {
+		e.log.Error(err, "failed to list FAR", "FAR uid", uid)
+		return nil, err
+	}
+
+	for _, far := range farList.Items {
+		if far.UID == uid {
+			return &far, nil
+		}
+	}
+
+	err := fmt.Errorf("could not find any FAR matching the UID")
+	e.log.Error(err, "failed to get far", "uid", uid)
+
+	return nil, err
+}
+
+func (e *executer) updateStatus(far *v1alpha1.FenceAgentsRemediation, err error) error {
+	var (
+		status          metav1.ConditionStatus
+		reason, message string
+	)
+	if err == nil {
+		status = metav1.ConditionTrue
+		reason = string(v1alpha1.FenceAgentSucceeded)
+		message = v1alpha1.FenceAgentSucceededConditionMessage
 	} else {
-		e.log.Info("Command has been executed successfully", "stdout", stdoutBuf.String())
+		status = metav1.ConditionFalse
+		reason = "FenceAgentFailed"
+		message = "The fence agent command failed"
 	}
-	return stdoutBuf.String(), stderrBuf.String(), err
+
+	e.log.Info("Updating Status Condition", "fenceAgentActionSucceededConditionStatus", status, "reason", reason, "LastUpdateTime", far.Status.LastUpdateTime)
+
+	meta.SetStatusCondition(&far.Status.Conditions, metav1.Condition{
+		Type:    v1alpha1.FenceAgentActionSucceededType,
+		Status:  status,
+		Reason:  reason,
+		Message: message,
+	})
+
+	return e.Status().Update(e.ctx, far)
```

Comment:
Maybe we can update the status with [`updateConditions`](https://github.com/medik8s/fence-agents-remediation/blob/main/controllers/fenceagentsremediation_controller.go#L285-L373)  when there is no error?

Reply:
at first, I didn't want to export the function, but it would be a good plus having all the update centralized, so I'd say either use it for both outcomes (success and error), or none